flutter-idea/src/io/flutter/analytics/FlutterAnalysisServerListener.java

```diff
@@ -0,0 +1,520 @@
+  static final String ACCEPTED_COMPLETION = "acceptedCompletion";
+  static final String REJECTED_COMPLETION = "rejectedCompletion";
+  static final String E2E_IJ_COMPLETION_TIME = "e2eIJCompletionTime";
+  static final String DURATION = "duration";
+  static final String FAILURE = "failure";
+  static final String SUCCESS = "success";
+  static final String ERROR_TYPE_REQUEST = "R";
+  static final String ERROR_TYPE_SERVER = "@";
+
+  static final String DAS_STATUS_EVENT_TYPE = "analysisServerStatus";
+  static final String IS_BLAZE_SYNC_SUCCESSFUL = "isBlazeSyncSuccessful";
+  static final String IS_DART_ACTIVE_LANG = "isDartActiveLang";
+  static final String IS_DART_ONLY_ACTIVE_LANG = "isDartOnlyActiveLang";
+  static final String IS_DART_WORKSPACE_TYPE = "isDartWorkspaceType";
+  static final String IS_DERIVE_TARGETS_ENABLED = "isDeriveTargetsEnabled";
+  static final String[] ERROR_TYPES =
+    new String[] {
+      AnalysisErrorType.CHECKED_MODE_COMPILE_TIME_ERROR,
+      AnalysisErrorType.COMPILE_TIME_ERROR,
+      AnalysisErrorType.HINT,
+      AnalysisErrorType.LINT,
+      AnalysisErrorType.STATIC_TYPE_WARNING,
+      AnalysisErrorType.STATIC_WARNING,
+      AnalysisErrorType.SYNTACTIC_ERROR
+    };
+
+  static final String LOG_ENTRY_KIND = "kind";
+  static final String LOG_ENTRY_TIME = "time";
+  static final String LOG_ENTRY_DATA = "data";
+  static final String LOG_ENTRY_SDK_VERSION = "sdkVersion";
+
+  // variables to throttle certain, high frequency events
+  private static final int COMPUTED_ERROR_SAMPLE_RATE = 100;
+  private static final Duration INTERVAL_TO_REPORT_MEMORY_USAGE = Duration.ofHours(1);
+  final RequestListener requestListener;
+  final ResponseListener responseListener;
+  final DartQuickFixListener quickFixListener;
+  // instance members
+  private final Project project;
+  //private final LoggingService loggingService;
+  private final Map<String, List<AnalysisError>> pathToErrors;
+  private final Map<String, Instant> pathToErrorTimestamps;
+  private final Map<String, Instant> pathToHighlightTimestamps;
+  private final Map<String, Instant> pathToOutlineTimestamps;
+  private final Map<String, RequestDetails> requestToDetails;
+  FileEditorManagerListener fileEditorManagerListener;
+  LookupSelectionHandler lookupSelectionHandler;
+  private int computedErrorCounter = 0;
+  private Instant nextMemoryUsageLoggedInstant = Instant.EPOCH;
+
+  FlutterAnalysisServerListener(@NotNull Project project, DartAnalysisServerService analysisServer) {
+    this.project = project;
+    this.pathToErrors = new HashMap<>();
+    this.pathToErrorTimestamps = new HashMap<>();
+    this.pathToHighlightTimestamps = new HashMap<>();
+    this.pathToOutlineTimestamps = new HashMap<>();
+    this.requestToDetails = new HashMap<>();
+    try {
+      LookupManager.getInstance(project).addPropertyChangeListener(this::onPropertyChange);
+    } catch (NullPointerException e) {
```

Comment:
what object is null? Prefer to avoid NPE rather than ever catching them.

Reply:
`LookupManager.getInstance(project)` could be null. I think that only happens during unit tests and it would be better to explicitly check if we're running tests. I'll change it.